File: backend/app/schemas/executive.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class ExecutiveEntitySummary(BaseModel):
    id: str
    name: str
    potential_savings: Decimal
    variance_percent: Decimal
    opportunity_count: int


class ExecutiveSummaryResponse(BaseModel):
    as_of_date: datetime
    top_plant: ExecutiveEntitySummary | None
    top_product: ExecutiveEntitySummary | None
    top_category: ExecutiveEntitySummary | None
    top_component: ExecutiveEntitySummary | None
    total_potential_savings: Decimal
    opportunity_count: int


class QuickWin(BaseModel):
    rank: int
    title: str
    potential_savings: Decimal
    ease: str
    confidence: Decimal
    urgency: str
    why_now: str
    opportunity_id: str


class QuickWinsResponse(BaseModel):
    as_of_date: datetime
    items: list[QuickWin]


class PlantExecutiveItem(BaseModel):
    plant_id: str
    plant_code: str
    plant_name: str
    country: str
    unit_cost: Decimal
    variance_percent: Decimal
    potential_savings: Decimal
    attention_level: str
    primary_driver: str | None
    benchmark_status: str
    opportunity_count: int


class PlantsExecutiveResponse(BaseModel):
    as_of_date: datetime
    items: list[PlantExecutiveItem]


class ProductExecutiveItem(BaseModel):
    product_id: str
    product_name: str
    equipment_family: str
    category_id: str
    category_name: str
    average_unit_cost: Decimal
    highest_cost_plant: str | None
    lowest_cost_plant: str | None
    variance_percent: Decimal
    potential_savings: Decimal
    priority: str
    opportunity_count: int


class ProductsExecutiveResponse(BaseModel):
    as_of_date: datetime
    attribution_policy: str = Field(
        default="PRIMARY_COMPATIBLE_MODEL",
        description="Each part is assigned to one stable equipment model for portfolio aggregation.",
    )
    items: list[ProductExecutiveItem]


class ProductPlantItem(BaseModel):
    plant_id: str
    plant_code: str
    plant_name: str
    country: str
    unit_cost: Decimal
    benchmark_cost: Decimal
    variance_percent: Decimal
    potential_savings: Decimal
    opportunity_count: int


class ProductComponentItem(BaseModel):
    component_id: str
    component_name: str
    part_classification: str
    potential_savings: Decimal
    variance_percent: Decimal
    confidence_score: Decimal
    opportunity_count: int
    lead_opportunity_id: str


class ProductDetailResponse(BaseModel):
    as_of_date: datetime
    product_id: str
    product_name: str
    equipment_family: str
    category_id: str
    category_name: str
    average_unit_cost: Decimal
    benchmark_unit_cost: Decimal
    variance_amount: Decimal
    variance_percent: Decimal
    annual_volume: int
    annual_spend: Decimal
    potential_savings: Decimal
    confidence_score: Decimal
    priority: str
    opportunity_count: int
    lead_opportunity_id: str
    snapshot_at: datetime
    highest_cost_plant: str | None
    lowest_cost_plant: str | None
    plants: list[ProductPlantItem]
    components: list[ProductComponentItem]


class ProductTrendPoint(BaseModel):
    period_start: date
    unit_cost: Decimal
    benchmark_cost: Decimal


class ProductTrendSeries(BaseModel):
    plant_id: str
    plant_name: str
    points: list[ProductTrendPoint]


class ProductTrendResponse(BaseModel):
    as_of_date: datetime
    product_id: str
    series: list[ProductTrendSeries]


class ComponentSupplierItem(BaseModel):
    supplier_id: str
    supplier_name: str
    country: str
    is_primary_supplier: bool
    unit_cost: Decimal
    annual_volume: int
    annual_spend: Decimal
    quality_score: Decimal
    delivery_score: Decimal
    overall_score: Decimal


class ComponentProductItem(BaseModel):
    product_id: str
    product_name: str


class ComponentDetailResponse(BaseModel):
    as_of_date: datetime
    component_id: str
    component_name: str
    part_classification: str
    annual_opportunity: Decimal
    annual_spend: Decimal
    annual_volume: int
    volume_change_percent: Decimal
    confidence_score: Decimal
    variance_percent: Decimal
    annual_tariff_impact: Decimal
    tariff_per_unit: Decimal
    duty_rate: Decimal
    peer_duty_rate: Decimal
    hs_code: str
    opportunity_count: int
    lead_opportunity_id: str
    lead_plant: str
    priority: str
    current_supplier: str | None
    benchmark_supplier: str | None
    current_supplier_unit_cost: Decimal | None
    benchmark_supplier_unit_cost: Decimal | None
    supplier_delta_percent: Decimal | None
    commercial_delta: Decimal | None
    suppliers: list[ComponentSupplierItem]
    products: list[ComponentProductItem]


class EquipmentCategoryItem(BaseModel):
    category_id: str
    category_name: str
    product_count: int
    annual_spend: Decimal
    potential_savings: Decimal
    cost_variance_percent: Decimal
    high_priority_opportunities: int
    confidence: Decimal
    primary_opportunity_driver: str | None
    priority: str


class EquipmentCategoriesResponse(BaseModel):
    as_of_date: datetime
    categories: list[EquipmentCategoryItem]


class CostDriverItem(BaseModel):
    driver_code: str
    driver_name: str
    benchmark_cost: Decimal
    comparison_cost: Decimal
    gap: Decimal
    contribution_percent: Decimal


class CostDriversResponse(BaseModel):
    as_of_date: datetime
    product_id: str
    overall_gap: Decimal
    contribution_total: Decimal
    drivers: list[CostDriverItem]


class ExecutiveReportResponse(BaseModel):
    as_of_date: datetime
    period: str | None
    scope: str
    summary: ExecutiveSummaryResponse
    plants: list[PlantExecutiveItem]
    products: list[ProductExecutiveItem]
    categories: list[EquipmentCategoryItem]
    quick_wins: list[QuickWin]

File: frontend/src/types/executive.ts
export interface ExecutiveEntitySummary {
  id: string;
  name: string;
  potential_savings: number;
  variance_percent: number;
  opportunity_count: number;
}

export interface ExecutiveSummary {
  as_of_date: string;
  top_plant: ExecutiveEntitySummary | null;
  top_product: ExecutiveEntitySummary | null;
  top_category: ExecutiveEntitySummary | null;
  top_component: ExecutiveEntitySummary | null;
  total_potential_savings: number;
  opportunity_count: number;
}

export interface QuickWin {
  rank: number;
  title: string;
  potential_savings: number;
  ease: "HIGH" | "MEDIUM" | "LOW";
  confidence: number;
  urgency: "HIGH" | "MEDIUM" | "LOW";
  why_now: string;
  opportunity_id: string;
}

export interface PlantExecutiveItem {
  plant_id: string;
  plant_code: string;
  plant_name: string;
  country: string;
  unit_cost: number;
  variance_percent: number;
  potential_savings: number;
  attention_level: "HIGH" | "WATCH" | "STABLE";
  primary_driver: string | null;
  benchmark_status: string;
  opportunity_count: number;
}

export interface ProductExecutiveItem {
  product_id: string;
  product_name: string;
  equipment_family: string;
  average_unit_cost: number;
  highest_cost_plant: string | null;
  lowest_cost_plant: string | null;
  variance_percent: number;
  potential_savings: number;
  priority: "HIGH" | "MEDIUM" | "LOW";
  opportunity_count: number;
}

export interface CategoryExecutiveItem {
  category_code: string;
  category: string;
  benchmark_cost: number;
  comparison_cost: number;
  gap: number;
  contribution_percent: number;
}

export interface CategoriesResponse extends DatedList<CategoryExecutiveItem> {
  overall_gap: number;
  contribution_total: number;
}

export interface ExecutiveAssistantReply {
  answer: string;
  session_id: string;
  model: string;
  provider: "vertex_ai" | "local_grounded_fallback";
  provider_note?: string | null;
  sources: { type: string; label: string }[];
}

export interface DatedList<T> { as_of_date: string; items: T[] }
export interface ProductsResponse extends DatedList<ProductExecutiveItem> { attribution_policy: string }

export interface ProductPlantItem {
  plant_id: string;
  plant_code: string;
  plant_name: string;
  country: string;
  unit_cost: number;
  benchmark_cost: number;
  variance_percent: number;
  potential_savings: number;
  opportunity_count: number;
}

export interface ProductComponentItem {
  component_id: string;
  component_name: string;
  category: string;
  potential_savings: number;
  variance_percent: number;
  confidence_score: number;
  opportunity_count: number;
  lead_opportunity_id: string;
}

export interface ProductDetailResponse {
  as_of_date: string;
  product_id: string;
  product_name: string;
  equipment_family: string;
  average_unit_cost: number;
  benchmark_unit_cost: number;
  variance_amount: number;
  variance_percent: number;
  annual_volume: number;
  annual_spend: number;
  potential_savings: number;
  confidence_score: number;
  priority: "HIGH" | "MEDIUM" | "LOW";
  opportunity_count: number;
  lead_opportunity_id: string;
  snapshot_at: string;
  highest_cost_plant: string | null;
  lowest_cost_plant: string | null;
  plants: ProductPlantItem[];
  components: ProductComponentItem[];
}

export interface ProductTrendPoint {
  period_start: string;
  unit_cost: number;
  benchmark_cost: number;
}

export interface ProductTrendSeries {
  plant_id: string;
  plant_name: string;
  points: ProductTrendPoint[];
}

export interface ProductTrendResponse {
  as_of_date: string;
  product_id: string;
  series: ProductTrendSeries[];
}

export interface ComponentSupplierItem {
  supplier_id: string;
  supplier_name: string;
  country: string;
  is_primary_supplier: boolean;
  unit_cost: number;
  annual_volume: number;
  annual_spend: number;
  quality_score: number;
  delivery_score: number;
  overall_score: number;
}

export interface ComponentDetailResponse {
  as_of_date: string;
  component_id: string;
  component_name: string;
  category: string;
  annual_opportunity: number;
  annual_spend: number;
  annual_volume: number;
  volume_change_percent: number;
  confidence_score: number;
  variance_percent: number;
  annual_tariff_impact: number;
  tariff_per_unit: number;
  duty_rate: number;
  peer_duty_rate: number;
  hs_code: string;
  opportunity_count: number;
  lead_opportunity_id: string;
  lead_plant: string;
  priority: "HIGH" | "MEDIUM" | "LOW";
  current_supplier: string | null;
  benchmark_supplier: string | null;
  current_supplier_unit_cost: number | null;
  benchmark_supplier_unit_cost: number | null;
  supplier_delta_percent: number | null;
  commercial_delta: number | null;
  suppliers: ComponentSupplierItem[];
  products: { product_id: string; product_name: string }[];
}

export interface GeneratedExecutiveReport {
  as_of_date: string;
  period: string;
  scope: string;
  narrative: string;
  provider: "vertex_ai" | "local_grounded_fallback";
  provider_note?: string | null;
  model: string;
  session_id: string;
  product_id?: string | null;
  file_name: string;
  report_id: string;
  download_url: string;
  storage: "minio";
}
